crates/context/config/src/client/protocol/starknet.rs

```diff
@@ -312,9 +313,76 @@ impl Network {
             }])
             .send()
             .await
-            .unwrap();
+            .map_err(|e| StarknetError::Custom {
+                operation: ErrorOperation::Mutate,
+                reason: format!("Failed to send transaction: {}", e),
+            })?;
+
+        let sent_at = Instant::now();
+        let timeout = Duration::from_secs(60); // Same 60-second timeout as NEAR
+
+        let receipt = loop {
+            match account
+                .provider()
+                .get_transaction_receipt(response.transaction_hash)
+                .await
+            {
+                Ok(receipt) => {
+                    if let starknet::core::types::TransactionReceipt::Invoke(invoke_receipt) =
+                        &receipt.receipt
+                    {
+                        if matches!(
+                            invoke_receipt.finality_status,
+                            TransactionFinalityStatus::AcceptedOnL2
+                                | TransactionFinalityStatus::AcceptedOnL1
+                        ) {
+                            break receipt;
+                        }
+
+                        if sent_at.elapsed() > timeout {
+                            return Err(StarknetError::TransactionTimeout);
+                        }
+                        continue;
+                    }
+                }
+                Err(err) => {
+                    return Err(StarknetError::Custom {
+                        operation: ErrorOperation::Mutate,
+                        reason: err.to_string(),
+                    });
+                }
+            }
+        };
 
-        let transaction_hash: Vec<u8> = vec![response.transaction_hash.to_bytes_be()[0]];
-        Ok(transaction_hash)
+        // Process the receipt
+        match receipt.receipt {
+            starknet::core::types::TransactionReceipt::Invoke(invoke_receipt) => {
+                match invoke_receipt.execution_result {
+                    ExecutionResult::Succeeded => {
+                        // Process events and return result
+                        for event in invoke_receipt.events.iter() {
+                            if event.from_address == contract_id {
```

Comment:
again, this event thing, not ideal. did you find an alternative? something along the lines of progressively awaiting the chain of execution until all x-calls are done?

Reply:
What do you mean? 
The idea was that we wait for receipt to be updated and then check the emitted events. 
This is the transaction lifecycle: https://docs.starknet.io/architecture-and-concepts/network-architecture/transaction-life-cycle/#transaction_status

Reply:
Yes, added checks for this specific event